src/main/kotlin/org/jitsi/nlj/rtcp/KeyframeRequester.kt

```diff
@@ -43,37 +43,71 @@ class KeyframeRequester : TransformerNode("Keyframe Requester") {
-
-        var drop = false
-        for (pkt in compoundRtcp.packets)
-        {
-            if (pkt !is RtcpFbFirPacket || pkt !is RtcpFbPliPacket) {
-                continue
-            }
-
-            val firCommandSeqNum = generateFirCommandSequenceNumber(pkt.mediaSourceSsrc, System.currentTimeMillis())
-            if (firCommandSequenceNumber < 0)
-            {
-                // NOTE(george): dropping the packet here should work fine as long as we don't
-                // receive 2 RTCP packets: one we want to drop and one we want to forward in the
-                // same compound packet.
-                drop = true
+        val packet = packetInfo.packet
+        val pliOrFir = when {
+            packet is CompoundRtcpPacket -> {
+                packet.packets.first { it is RtcpFbPliPacket || it is RtcpFbFirPacket }
             }
-            else if ((pkt is RtcpFbPliPacket && !hasPliSupport) || (pkt is RtcpFbFirPacket) && !hasFirSupport)
-            {
-                // NOTE(george): same as above
-                drop = true
+            packet is RtcpFbFirPacket -> packet
+            packet is RtcpFbPliPacket -> packet
+            else -> null
+        }
 
-                requestKeyframeInternal(pkt.mediaSenderSsrc, firCommandSequenceNumber)
+        var forward = true
+        when {
+            pliOrFir is RtcpFbPliPacket -> {
+                when {
+                    hasPliSupport -> {
+                        // Forward PLI
+                        forward = forwardPliOrFir(pliOrFir)
+                    }
+                    hasFirSupport -> {
+                        // Translate to FIR
+                        requestKeyframe(pliOrFir.mediaSenderSsrc)
+                        forward = false
+                    }
+                    else -> {
+                        // Can't do anything
+                        forward = false
+                    }
+                }
             }
-            else if (pkt is RtcpFbFirPacket)
-            {
-                RtcpFbFirPacket.setSeqNum(pkt.buffer, pkt.offset, firCommandSeqNum)
+            pliOrFir is RtcpFbFirPacket -> {
+                when {
+                    hasPliSupport -> {
+                        // Translate to PLI
+                        requestKeyframe(pliOrFir.mediaSenderSsrc)
+                        forward = false
+                    }
+                    hasFirSupport -> {
+                        // Forward FIR
+                        forward = forwardPliOrFir(pliOrFir)
+                    }
+                    else -> {
+                        // Can't do anything
+                        forward = false
+                    }
+                }
             }
         }
 
-        return if (drop) null else packetInfo
+        return if (forward) packetInfo else null
+    }
+
+    private fun forwardPliOrFir(pliOrFir: RtcpFbPacket): Boolean {
+        val firCommandSeqNum =
```

Comment:
it seems weird to me that we generate a FIR command seq num even if it's not a FIR packet we're forwarding.  since PLI requires no other special handling (so this method does nothing for PLIs), maybe we should just do this logic for FIR in the FIR blocks and call nothing for PLI? so:

```
       when {
            pliOrFir is RtcpFbPliPacket -> {
                when {
                    hasPliSupport -> {
                        // Forward PLI 
                        forward = true //BRIAN: just say we're going to forward here
                    }
                    hasFirSupport -> {
                        // Translate to FIR
                        requestKeyframe(pliOrFir.mediaSenderSsrc)
                        forward = false
                    }
                    else -> {
                        // Can't do anything
                        forward = false
                    }
                }
            }
pliOrFir is RtcpFbFirPacket -> {
            when {
                hasPliSupport -> {
                        // Translate to PLI
                        requestKeyframe(pliOrFir.mediaSenderSsrc)
                        forward = false
                    }
                    hasFirSupport -> {
                        // Forward FIR
                        //BRIAN: just do the FIR-specific work in place here
                       val firSeqNum = generateFirCommandSequenceNumber(pliOrFir.mediaSourceSsrc, System.currentTimeMillis())
                       forward = if (firSeqNum >= 0) {
                            pliOrFir.seqNum = firCommandSeqNum
                            true
                       } else {
                            false
                       }
                    }
                    else -> {
                        // Can't do anything
                        forward = false
                    }
                }
            }
        }
```